rollup/internal/controller/relayer/l2_relayer_sanity.go

```diff
@@ -0,0 +1,271 @@
+	if err != nil {
+		return err
+	}
+
+	// Validate each batch in detail
+	if err := r.validateBatchesDetailed(batchesToSubmit, prevChunk); err != nil {
+		return err
+	}
+
+	log.Info("Sanity check passed before constructing transaction", "batches count", len(batchesToSubmit))
+	return nil
+}
+
+// validateBatchesBasic performs basic validation on all batches
+func (r *Layer2Relayer) validateBatchesBasic(batchesToSubmit []*dbBatchWithChunks) error {
+	for i, batch := range batchesToSubmit {
+		if batch == nil || batch.Batch == nil {
+			return fmt.Errorf("batch %d is nil", i)
+		}
+
+		if len(batch.Chunks) == 0 {
+			return fmt.Errorf("batch %d has no chunks", batch.Batch.Index)
+		}
+	}
+	return nil
+}
+
+// validateCodecVersions checks all batches have the same codec version
+func (r *Layer2Relayer) validateCodecVersions(batchesToSubmit []*dbBatchWithChunks) error {
+	firstBatchCodecVersion := batchesToSubmit[0].Batch.CodecVersion
+	for _, batch := range batchesToSubmit {
+		if batch.Batch.CodecVersion != firstBatchCodecVersion {
+			return fmt.Errorf("batch %d has different codec version %d, expected %d", batch.Batch.Index, batch.Batch.CodecVersion, firstBatchCodecVersion)
+		}
+	}
+	return nil
+}
+
+// getPreviousChunkForContinuity gets the previous chunk for block continuity check
+func (r *Layer2Relayer) getPreviousChunkForContinuity(firstBatch *dbBatchWithChunks) (*orm.Chunk, error) {
+	firstChunk := firstBatch.Chunks[0]
+	if firstChunk.Index == 0 {
+		return nil, fmt.Errorf("genesis chunk should not be in normal batch submission flow, chunk index: %d", firstChunk.Index)
+	}
+
+	prevChunk, err := r.chunkOrm.GetChunkByIndex(r.ctx, firstChunk.Index-1)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get previous chunk %d for continuity check: %w", firstChunk.Index-1, err)
+	}
+
+	return prevChunk, nil
+}
+
+// validateBatchesDetailed performs detailed validation on each batch
+func (r *Layer2Relayer) validateBatchesDetailed(batchesToSubmit []*dbBatchWithChunks, prevChunkFromPrevBatch *orm.Chunk) error {
+	for i, batch := range batchesToSubmit {
+		if err := r.validateSingleBatch(batch, i, batchesToSubmit, prevChunkFromPrevBatch); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// validateSingleBatch validates a single batch and its chunks
+func (r *Layer2Relayer) validateSingleBatch(batch *dbBatchWithChunks, i int, allBatches []*dbBatchWithChunks, prevChunkFromPrevBatch *orm.Chunk) error {
+	// Validate batch fields
+	if err := r.validateBatchFields(batch, i, allBatches); err != nil {
+		return err
+	}
+
+	// Validate message queue consistency
+	if err := r.validateMessageQueueConsistency(batch.Batch.Index, batch.Chunks, common.HexToHash(batch.Batch.PrevL1MessageQueueHash), common.HexToHash(batch.Batch.PostL1MessageQueueHash)); err != nil {
+		return err
+	}
+
+	// Validate chunks
+	if err := r.validateBatchChunks(batch, i, allBatches, prevChunkFromPrevBatch); err != nil {
```

Comment:
why not pass the correct `lastChunkFromPrevBatch` here instead of doing it inside the function?